GCP: Use CAM Deployment Service Account

Edwin-Pau opened the pull request:
To support autoscaling of CAC's in the future, new CAC tokens need
to be created during CAC installations. To generate new connector
tokens, the CAM python script is used during provisioning of the
connector, which makes CAM API calls using the user-provided CAM
deployment-level service account credentials.

Both plaintext and encrypted secrets deployments are supported using
CAM deployment-level service accounts instead of connector tokens.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Use CAM deployment service account (single)

To support autoscaling of CAC's in the future, new CAC tokens need
to be created during CAC installations. To generate new
CAC tokens, the CAM python script is used during provisioning of the
connector, which makes CAM API calls using the user-provided CAM
deployment-level service account credentials.

Both plaintext and encrypted secrets deployments are supported using
CAM deployment-level service accounts instead of CAC tokens.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Add CAM python script for connector use

Added a cac-cam.py script to be used for both single-connector and
multi-region GCP deployment types. This python script is used by
the CAC provisioning scripts so that it can use a CAM deployment-
level service account (JSON credentials file) to retrieve a new
connector token during CAC installations.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Use CAM deployment service account (multi)

To support autoscaling of CAC's in the future, new CAC tokens need
to be created during CAC installations. To generate new connector
tokens, the CAM python script is used during provisioning of the
connector, which makes CAM API calls using the user-provided CAM
deployment-level service account credentials.

Both plaintext and encrypted secrets deployments are supported using
CAM deployment-level service accounts instead of connector tokens.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Update README CAM deployment service account

Updated the documentation since the GCP deployments (quickstart,
single-connector and multi-region) now use a CAM deployment-level
service account credentials file to create a new connector token string
prior to installing the connector.

Added instructions on how to obtain CAM deployment-level service account
credentials, as well as instructions to use CAM deployment-level
service account file in terraform.tfvars.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Add function to create CAM deployment SA (QS)

Added a function in the quickstart CAM script to create a deployment-
level service account using the user-provided auth_token. This function
uses the endpoint for creating an API key, which returns the username,
apiKey, and tenantId.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Use CAM deployment service account (QS)

Modified the GCP quickstart script so that it works with the CAM
deployment-level service account change. Instead of creating a connector
token using the API token, it now creates and uses the CAM
deployment-level service account and passed to the GCP single-connector
deployment that quickstart uses.

Signed-off-by: Edwin-Pau <epau@teradici.com>

shermanyin commented on modules/gcp/cac-igm/cac-cam.py:
```diff
@@ -0,0 +1,149 @@
+#!/usr/bin/python3
```
best to use #!/usr/bin/env python3

shermanyin commented on modules/gcp/cac-igm/cac-cam.py:
```diff
@@ -0,0 +1,149 @@
+
+    request_body = dict(username = cam_credentials.get('username'), 
+                        password = cam_credentials.get('apiKey'),
+                        tenantId = cam_credentials.get('tenantId'))
```
don't think you need tenantId

shermanyin commented on deployments/gcp/single-connector/terraform.tfvars.sample:
```diff
@@ -1,6 +1,6 @@
 # Commented out lines represents defaults that can be changed
 # On Windows systems, the default backslash \ path separator must be changed to forward slash for any path variables.
-# Example: gcp_credentials_file = "C:/path/to/cred.json"
+# Example: gcp_credentials_file = "C:/path/to/gcp_cred.json"
```
why the change? if we call it gcp_cred then we should change line 5 too?

shermanyin commented on deployments/gcp/single-connector/vars.tf:
```diff
@@ -159,8 +159,13 @@ variable "ws_subnet_cidr" {
+}
+
+variable "cam_deployment_sa_file" {
+  description = "Location of CAM deployment-level service account JSON file"
```
our docs call it "Deployment Service Account", so let's stick with that name. Same with other places in your commits.

shermanyin commented on modules/gcp/cac/cac-provisioning.sh.tmpl:
```diff
@@ -24,20 +24,26 @@ log() {
+install_python() {
+    log "--> Installing Python and Python3..."
+    apt-get -qq update
+    apt install -y python
```
is this required? can we just install python3 by calling python3 instead of python in line 53?

shermanyin commented on modules/gcp/cac/main.tf:
```diff
@@ -6,12 +6,32 @@
+
+  bucket  = var.bucket_name
+  name   = local.cam_script
+  source = "${path.module}/cac-cam.py"
```
use local.cam_script here

shermanyin commented on modules/gcp/cac/cac-provisioning.sh.tmpl:
```diff
@@ -52,10 +58,29 @@ get_credentials() {
-        log "--> Decrypting CAC token..."
-        data=$(echo "{ \"ciphertext\": \"${cac_token}\" }")
+        log "--> Decrypting CAM deployment-level service account JSON file..."
+        cam_cred_encrypted=$(cat ${cam_deployment_sa_file})
```
We downloaded to $INSTALL_DIR, so best reference $INSTALL_DIR/${cam_deployment_sa_file}, even though if it's redundant now (in case we change where we cd into in the future). Same for the rest of this file.

shermanyin commented on modules/gcp/cac/cac-provisioning.sh.tmpl:
```diff
@@ -24,20 +24,26 @@ log() {
+
 get_credentials() {
+    log "--> Downloading CAM deployment-level service account JSON file from the bucket..."
+    gsutil cp gs://${bucket_name}/${cam_deployment_sa_file} $INSTALL_DIR
```
put this together in the section where you handle decrypting of the file?

shermanyin commented on modules/gcp/cac/cac-provisioning.sh.tmpl:
```diff
@@ -52,10 +58,29 @@ get_credentials() {
+
+    log "--> Downloading CAM python script from the bucket..."
+    gsutil cp gs://${bucket_name}/${cam_script} $INSTALL_DIR
+    chmod +x ${cam_script}
```
$INSTALL_DIR

shermanyin commented on modules/gcp/cac/cac-provisioning.sh.tmpl:
```diff
@@ -52,10 +58,29 @@ get_credentials() {
+    chmod +x ${cam_script}
+
+    # Set CAC_TOKEN variable using the script's output
+    CAC_TOKEN=`./${cam_script} ${cam_deployment_sa_file}`
```
$INSTALL_DIR

shermanyin commented on quickstart/cam.py:
```diff
@@ -48,6 +48,21 @@ def deployment_add_gcp_account(self, key, deployment):
         resp.raise_for_status()
 
+    def deployment_key_create(self, deployment):
+        deployment_id = {
```
if you follow the conventions in other functions, then it would be key_details

shermanyin replied:
provide option to choose keyName

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -242,6 +243,15 @@ def ssh_key_create(path):
     subprocess.run(ssh_cmd.split(' '), check=True)
 
 
+def cam_deployment_sa_create_key(service_account, filepath):
```
this is just writing a dictionary to a file, don't think we need a custom function for just 2 lines.

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -326,8 +336,16 @@ def kms_python_client_install():
+    cam_deployment_key = mycam.deployment_key_create(deployment)
+
+    print('Building CAM deployment service account...')
+    cam_deployment_sa = {
```
not sure why we need to do this, you're just taking out the keyName? shouldn't be an issue if you left it in and just use cam_deployment_key

shermanyin commented on quickstart/gcp-cloudshell-quickstart.py:
```diff
@@ -51,6 +51,7 @@
 SECRETS_DIR      = 'secrets'
 SA_KEY_PATH      = SECRETS_DIR + '/gcp_service_account_key.json'
```
probably want to call this GCP_SA_KEY_PATH now that there are multiple SAs involved.

shermanyin reviewed (commented):
Please replace "deployment-level service account" with "Deployment Service Account", and references to its key "Deployment Service Account Key".

Commit:
GCP: Fix shebang line for cac-cam.py scripts

Fixed the shebang line for the cac-cam.py scripts to use /usr/bin/env
which uses the user's path to run the script, making it more portable
to different environments.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Fix cac-cam.py scripts

Removed tenantId from request_body when retrieving authentication token.

Fixed variable reference when raising an error. Inserted newline
between try and except blocks.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Rename to CAM Deployment Service Account

Renamed all references and names used from CAM deployment-level
service account to CAM Deployment Service Account.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Update terraform.tfvars.sample files

Updated all 3 GCP deployment types terraform.tfvars.sample files. The
GCP credentials file reference has been updated to gcp_cred.json to be
more explicit since there are multiple service accounts in the tfvars
file.

The CAM Deployment Service Account reference has been updated
to cloud-access-manager-service-account.json since this is the default
name file name when downloaded from CAM.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Fix CAC scripts and main.tf for CAM SA

Changed the script so that only Python3 is installed and used.

References to the downloaded files now use absolute paths instead of
the current path. IE. cam_deployment_sa_file, cam_script

Used a local variable instead of hard coding in CACs' main.tf files.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Fix quickstart changes for CAM SA

Removed unnecessary function for writing the key to a file.

Removed cam_deployment_sa dictionary.

Renamed variable to key_details to follow
convention.

Renamed GCP service account to GCP_SA_KEY_PATH since there
are now different service accounts used within the script.

Signed-off-by: Edwin-Pau <epau@teradici.com>

Commit:
GCP: Move dependencies from script to instance

The dependencies of several google_storage_bucket_object have been
moved from inside the provisioning script resource to inside the CAC
instance resource instead.

The dependency for the provisioning script has also been commented out,
since this should be inferred by Terraform when creating the instance.

Renamed the resource for the CAM credentials file to be
cam-deployment-sa-file so that it's the same for both cac and cac-igm.

Signed-off-by: Edwin-Pau <epau@teradici.com>
